aspects.bzl

```diff
@@ -0,0 +1,228 @@
+
+rust_analyzer will generate a rust-project.json file for the
+given targets. This file can be consumed by rust-analyzer as an alternative
+to Cargo.toml files.
+"""
+
+load("@io_bazel_rules_rust//rust:private/utils.bzl", "find_toolchain")
+
+_rust_rules = [
+    "rust_library",
+    "rust_binary",
+]
+
+TargetInfo = provider(
+    fields = {
+        'name' : 'target name',
+        'root' : 'crate root',
+        'edition' : 'edition',
+        'dependencies' : 'dependencies',
+        'transitive_deps' : 'closure of all transitive dependencies',
+        'cfgs' : 'compilation cfgs',
+        'env' : 'Environment variables, used for the `env!` macro',
+    }
+)
+
+def _rust_project_aspect_impl(target, ctx):
+  # We support only these rule kinds.
+  if ctx.rule.kind not in _rust_rules:
+    return []
+
+  info = ctx.toolchains["@io_bazel_rules_rust//rust:toolchain"]
+
+  # extract the crate_root path
+  edition = ctx.rule.attr.edition
+  if not edition:
+    # TODO if edition isn't specified, default to what is in the rust toolchain
+    edition = info.default_edition
+
+  crate_name = ctx.rule.attr.name
+
+  crate_root = ctx.rule.attr.crate_root
+  if not crate_root:
+    if len(ctx.rule.attr.srcs) == 1:
+      crate_root = ctx.rule.attr.srcs[0]
+    else:
+      for src in ctx.rule.attr.srcs:
+        file_name = src.label.name
+        crate_type = ctx.rule.attr.crate_type
+        if crate_type == "bin" and file_name.endswith("main.rs"):
+          crate_root = src
+          break
+        if crate_type == "rlib" or crate_type == "dylib":
+          if file_name.endswith("lib.rs"):
+            crate_root = src
+            break
+  # this will always be the first in the depset
+  crate_root = crate_root.files.to_list()[0].path
+
+  cfgs = []
+  for feature in ctx.rule.attr.crate_features:
+    cfgs.append("feature=\"" + feature + "\"")
+  for flag in ctx.rule.attr.rustc_flags:
+    # --cfg flags should be passed as well but as an atomic
+    # config, not key/value
+    if flag.startswith("--cfg"):
+      cfgs.append(flag[6:])
+  env = ctx.rule.attr.rustc_env
+
+  deps = []
+  transitive_deps = []
+  for dep in ctx.rule.attr.deps:
+    deps.append(dep[TargetInfo])
+    transitive_deps.append(dep[TargetInfo])
+    if TargetInfo not in dep:
+      continue
+    transitive_deps.extend(dep[TargetInfo].dependencies)
+
+  return [TargetInfo(
+      name = crate_name,
+      edition = edition,
+      cfgs = cfgs,
+      root= crate_root,
+      env=env,
+      dependencies = deps,
+      transitive_deps = transitive_deps)]
+
+rust_project_aspect = aspect(
+    attr_aspects = ["deps"],
+    implementation = _rust_project_aspect_impl,
+    toolchains = [ "@io_bazel_rules_rust//rust:toolchain" ]
+)
+
+def create_crate(target):
+  crate = dict()
+  crate["name"] = target.name
+  crate["ID"] = "ID-" + target.name
+  crate["root_module"] = target.root
+  crate["edition"] = target.edition
+  # TODO(bwb): smarter heuristic for workpace member indexing
```

Comment:
What does that mean?

Reply:
This is a new field I haven't worked with before in the rust-project.json. Basically it creates a two tier indexing system - it'll re-index things faster if it's within the workspace. Sysroot is obviously not in the workspace, but things like 3rd party imports also likely shouldn't be marked as in the workspace.